assets/js/googlesitekit/data/create-notifications-store.js

```diff
@@ -0,0 +1,259 @@
+				type: RECEIVE_NOTIFICATIONS,
+			};
+		},
+
+		/**
+		 * Dispatches an action signifying the `fetchNotifications` side-effect failed.
+		 *
+		 * @since n.e.x.t
+		 * @private
+		 * @return {Object} Redux-style action.
+		 */
+		receiveNotificationsFailed() {
+			return {
+				payload: {},
+				type: RECEIVE_NOTIFICATIONS_FAILED,
+			};
+		},
+	};
+
+	const controls = {
+		[ FETCH_NOTIFICATIONS ]: () => {
+			return API.get( type, identifier, datapoint );
+		},
+	};
+
+	const reducer = ( state = INITIAL_STATE, action ) => {
+		switch ( action.type ) {
+			case ADD_NOTIFICATION: {
+				const { notification } = action.payload;
+
+				return {
+					...state,
+					clientNotifications: {
+						...state.clientNotifications,
+						[ notification.id ]: notification,
+					},
+				};
+			}
+
+			case REMOVE_NOTIFICATION: {
+				const { id } = action.payload;
+
+				// At this point, only client-side notifications can be removed.
+				if ( 'undefined' === typeof state.clientNotifications[ id ] ) {
+					// Trigger a warning clarifying that if a server-side notification is attempted to be removed.
+					if ( 'undefined' !== typeof state.serverNotifications &&
+						'undefined' !== typeof state.serverNotifications[ id ]
+					) {
+						global.console.warn( `Cannot remove server-side notification ${ id }; this may be changed in a future release.` );
+					}
+
+					return { ...state };
+				}
+
+				const newNotifications = { ...state.clientNotifications };
+				delete newNotifications[ id ];
+
+				return {
+					...state,
+					clientNotifications: newNotifications,
+				};
+			}
+
+			case FETCH_NOTIFICATIONS: {
+				return {
+					...state,
+					isFetchingNotifications: true,
+				};
+			}
+
+			case RECEIVE_NOTIFICATIONS: {
+				const { notifications } = action.payload;
+
+				return {
+					...state,
+					isFetchingNotifications: false,
+					serverNotifications: notifications.reduce(
+						( acc, notification ) => {
+							return {
+								...acc,
+								[ notification.id ]: notification,
+							};
+						},
+						{}
+					),
+				};
+			}
+
+			case RECEIVE_NOTIFICATIONS_FAILED: {
+				return {
+					...state,
+					isFetchingNotifications: false,
+				};
+			}
+
+			default: {
+				return { ...state };
+			}
+		}
+	};
+
+	const resolvers = {
+		*getNotifications() {
+			try {
+				const notifications = yield actions.fetchNotifications();
+				return actions.receiveNotifications( notifications );
+			} catch ( err ) {
+				return actions.receiveNotificationsFailed();
+			}
+		},
+	};
+
+	const selectors = {
+		/**
+		 * Gets the current notifications.
+		 *
+		 * Returns `undefined` if notifications are not available/loaded.
+		 *
+		 * @since n.e.x.t
+		 * @param {Object} state Data store's state.
+		 * @return {Array|undefined} Current list of notifications.
+		 */
+		getNotifications( state ) {
+			const { serverNotifications, clientNotifications } = state;
+
+			if ( 'undefined' === typeof serverNotifications ) {
```

Comment:
That's fair. Do you think it would make sense to return `undefined` if `serverNotifications` is `undefined` and `clientNotifications` is an empty object? At that point we could then at least fully rely on this knowing that notifications are still being loaded.

I would want to avoid seeing no notifications and suddenly seeing them afterwards.

Reply:
Yeah, good call. I can implement that while I handle the `console` spy test. I'd still consider that a "loading" state so that makes sense. 👍

Reply:
Circling back to this: I actually don't think that's the right approach. If the client notifications are removed I would expect the UI to reflect "empty" notifications rather than revert back to a "loading" state. If we want to show loading it might be better to have an `isDoingGetNotifications()` or something that returns `true` if server notifications are still loading and displays UI indicating that rather than having state go from "loading" to "not loading" to "loading" after adding/removing client notifications.

For now I opted to have this return an empty array, but if we want to not ever go from "empty state" to "showing notifications from server" state I think we need finer-grained selectors for the server loading state.